### DISCUSSION TITLE
Enhancement for multiline get property

### DISCUSSION THREAD

**Discussion by HakamFostok** [dotnet/csharplang#5715] [Language Ideas]
**Clarification: This Idea generated some confusion so I wanted to clarify**
**I am NOT doing the heavy calculation in the getter, when I need that I use methods** <br/>
**I am a big believer in the Clean Code principles, so the longest method I have is no more than 10 lines of code** <br/>
**The Idea here is to go with the same theme that C# has taken for the last couple of years of cleaning the ceremony and indentation**

<hr />

 If there is a get property with multi-line of code it should be written like this
this will cause the code inside the property to be intended for one more level
and we will have three more lines for the `get` keywords and for the opening and closing braces of it

    public IReadOnlyList<PlayerScore> Scores
    {
        get
        {
            // first line of code
           // second line of code

           // N line of code
        }
    }

Could it be like this?

    public get IReadOnlyList<PlayerScore> Scores
    {
        // first line of code
        // second line of code

        // N line of code
    }

this will save indentation and 3 lines in the file

Thank you

**Comment by Joe4evr** [comment]
Sounds more like you have a problem in your architecture and this should be a regular method instead. If you do so much computation in the getter body that the extra lines are bothering you, then using a property is the wrong choice.

**Reply by HakamFostok** [reply]
Thank you, <br/> 
I do not think I have a problem with the architecture, <br/> 
I like to use getter properties for small calculated properties that do not need parameters. <br/>
When there is heavy lifting I prefer methods, of course. <br/> 
I did not show the code inside the getter but most of my getters are no more than 3 lines of code and there is no heavy computation at all inside them<br/>
The bottom line that: I think this complies with the general theme that C# took for the last couple of years of making the boilerplate and ceremony less and saving vertical and horizontal indentation <br/>
I think this can be considered like the "File scoped namespaces" and "Implicit usings" enhancement that done with C# 10

**Reply by HaloFour** [reply]
IMO it's not worth having a special syntax for this purpose.  [Expression blocks](https://github.com/dotnet/csharplang/issues/3086) will allow you to exploit expression-bodied properties to achieve the same thing:

```cs
public IReadOnlyList<PlayerScore> Scores => {
    // first line of code
    // second line of code

    // N line of code
};
```

**Reply by HakamFostok** [reply]
@HaloFour  Thank you sir, So if I understand correctly, this is now is not valid code, but if the Expression blocks feature implemented then I will be able to write something like this, correct?

**Reply by HakamFostok** [reply]
I think I agree with you @HaloFour, I had no idea that there was an "Expression blocks" proposal, It serves the same thing that I wanted and I wish it will be implemented, in that case, there is no need for my suggestion, I agree.
Thank you again

**Reply by MgSam** [reply]
@Joe4evr I'm not sure making assumptions about people's code which you know nothing about is really productive. C# has supported getters with arbitrarily complex logic since V1.0. It's a little too late to complain about it being a bad design pattern.

**Comment by DavidArno** [comment]
This saves two of those three lines and avoids the indentation. And it can be written today:
```c#
public IReadOnlyList<PlayerScore> Scores => CalculateCurrentScores();
private IReadOnlyList<PlayerScore> CalculateCurrentScores()
{
    // first line of code
    // second line of code

    // N line of code
}
```

Deleting the below comment as it seems to be distracting the conversation away from the core point of my comment.
~~But as @Joe4evr says, really this indicates a problem with your class; multiline getters ought to be avoided if possible and you'd be better off making `CalculateCurrentScores`, public and getting rid of the getter.~~

**Reply by HakamFostok** [reply]
Thank you sir for your reply, but I think you missed my point completely, Please refer to the above comment that I wrote to clarify my point, Thank you

**Reply by DavidArno** [reply]
Surely your point is that you want to reduce the "boilerplate" around creating multi-line getters?

The approach I take on the odd occasion that I really need a multi-line getter is to create a private method and call that from the getter as per my example. This approach already works.

So, with that in mind, you want brand new syntax, with all the work that would entail, to save you just one line of code. That seems a big ask of the language team.

**Reply by HakamFostok** [reply]
@HaloFour has linked the Expression block proposal, I did not know about it before suggesting this, you are right I am asking for too much maybe, but if that feature is implemented then, of course, there is no need for my suggestion.
I hope it will be implemented, it will work for me perfectly <br/>
Thank you

**Comment by MgSam** [comment]
As @HaloFour already said, Expression Blocks would help solve this. 

More generally though, it's always seemed rather arbitrary to me that lambdas could use the arrow syntax for expressions or blocks but methods/getters/setters could only use arrow syntax for expressions. 

``` C#
var foo = () => "foo"; //OK
var bar = () => { return "bar"; };  //OK

string foo() => "foo"; //OK
string bar() => { return "bar"; } //Not OK
```

It makes the cliff for going from an expression-based getter-only property to one with a block quite annoying.
``` C#
class Example 
{
    string Foo => "foo"; //Nice
    string Bar { get { barHops++; return "bar"; } } //Ugly
    string Bar => { barHops++; return "bar"; } //Would be nice
    string Bar => { barHops++; "bar"; } //Would be even nicer - I believe Expression Blocks would allow this
    
    private int barHops = 0;
}
```

**Reply by CyrusNajmabadi** [reply]
> it's always seemed rather arbitrary to me that lambdas could use the arrow syntax for expressions or blocks 

The lambdas need this so that we actually know they're lambdas :)

Methods/props/getter/setters don't need this since we know waht they are.

**Comment by pleshw** [comment]
I liked the idea. The code is more in line with other modern languages, and its also easier to understand, explain and write.

**Comment by RikkiGibson** [comment]
I am mildly supportive of the idea. I think it could work by simply checking if the accessor list looks like it starts with an accessor, i.e. is it a sequence of attribute lists and modifiers followed by `get`/`set`/`init` and `{` or `=>` or `;`. If it doesn't then just assume it's a get accessor. I'd want to dig a bit further to be confident there's not a true ambiguity, though.

```cs
class Person
{
    public string FirstName { get; set; }
    public string LastName { get; set; }
    public string FullName
    {
        if (string.IsNullOrEmpty(FirstName)) return LastName;
        if (string.IsNullOrEmpty(LastName)) return FirstName;
        return FirstName + " " + LastName;
    }
}
```

I think this would make scenarios where methods are converted to properties less cumbersome. This also would not be unique to C#. As was pointed out, other prominent languages which have properties with accessor lists just treat the "property body" as a get accessor if no explicit accessor declarations are found within it.

All that said, the language team would need to be convinced that this would be an idea worth picking up in addition to "expression blocks". That seems doubtful 😅

**Reply by 333fred** [reply]
> All that said, the language team would need to be convinced that this would be an idea worth picking up in addition to "expression blocks". That seems doubtful sweat_smile

Agreed :slightly_smiling_face:

**Reply by CyrusNajmabadi** [reply]
@RikkiGibson i'm heavily opposed to this as it's going to get *super* confusing once we do "property scoped fields" (which was approved by LDM, but we ordered after 'semi-auto props').  With property scoped fields, you can write:

```c#
    public string FullName
    {
        int field1;
        bool field2;

        get { ... }
    }
```

The problem with allowing arbitrary statements is that now the first two lines are super ambiguous.  Are they field decls before more accessors?  Or are they local-statements before more statements? :(

**Reply by RikkiGibson** [reply]
That's a great point, thanks @CyrusNajmabadi.
